joomla/joomla-cms
# [5.0] Allow Joomla be served from a public* folder

Pull Request for Issue # .

### public folder == subfolder of the root installation (but not restricted to only this config) that has only the allowed entry points (index.php, administrator/index.php, api/index.php) and any public static files (ie `media/`, `images/`)

### Summary of Changes

- This is the actual code needed to get Joomla serving from a public folder. It doesn't include the code for the installation to produce/create the different serving option. Also it does't have the required symbolic links for both the `media` and `images` folders and neither the actual public folder (this is supposed to be generated)

- The purpose of this PR is to have people try to serve Joomla from a subfolder named `public` instead of the plain root that gives access to all the folders...

- The reason I opened this RFC is mainly because the actual changes required are minimal (that's since the introduction of the child templates)...

- Gotcha: legacy templates **WON'T** work with this!!!

- There is a new constant `JPATH_PUBLIC` which is used to conditionally change the code both in the HTMLHelper and the WAM so nothing is broken...


### Testing Instructions

- To try this you need to git clone Joomla
- install Joomla
- then apply this PR (check out this PR)
- then unzip in the root folder of Joomla this [public.zip](https://github.com/joomla/joomla-cms/files/11361929/public.zip)
- then symlink the root `media` to `{root}\public\media`
- also symlink the root `images` to `{root}\public\images`
- Point your server to the `public` folder (or if it's a local installation to `localhost\{root}\public` and you should still have a functioning Joomla instance, eg:
<img width="903" alt="Screenshot 2023-04-30 at 12 28 40" src="https://user-images.githubusercontent.com/3889375/235349896-223c6fdb-233b-43d2-80e5-d4ad808d0001.png">


Alternatively download and install from the link at the bottom of this PR (on Github) and
- then unzip in the root folder of Joomla this [public.zip](https://github.com/joomla/joomla-cms/files/11361929/public.zip)
- then symlink the root `media` to `{root}\public\media`, eg: `ln -s /path-to-joomla/media /path-to-joomla/public/media`
- also symlink the root `images` to `{root}\public\images`, eg: `ln -s /path-to-joomla/images /path-to-joomla/public/images `
- Point your server to the `public` folder (or if it's a local installation to `localhost\{root}\public` and you should still have a functioning Joomla instance
<img width="903" alt="Screenshot 2023-04-30 at 12 28 40" src="https://user-images.githubusercontent.com/3889375/235349896-223c6fdb-233b-43d2-80e5-d4ad808d0001.png">

### Actual result BEFORE applying this Pull Request

All good

### Expected result AFTER applying this Pull Request

All good but way safer

@HLeithner @laoneo
